py/htm/advanced/regions/ApicalTMPairRegion.py

```diff
@@ -522,3 +522,12 @@ def getAlgorithmInstance(self):
         Returns an instance of the underlying temporal memory instance
         """
         return self._tm
+
+    def saveConnectionsToFile(self, filepath):
+        """
+        Binary dumps connections objects into file specified, adding suffix specifying the type
+        """
+        with open(filepath+"_basal.dump", "wb") as f:
+            f.write(self._tm.basalConnections.save())
```

Comment:
nit: you could use the `connections.saveToFile("myfile.dump")`

Reply:
I can't, this is not exposed to python..

i tried just now adding to bindings py_Connections.cpp
`py_Connections.def("saveToFile", &Connections::saveToFile);`

But i got 
```
TypeError: saveToFile(): incompatible function arguments. The following argument types are supported:
    1. (self: htm.bindings.algorithms.Connections, arg0: str, arg1: htm.bindings.engine_internal.SerializableFormat) -> None

Invoked with: <htm.advanced.algorithms.connections.Connections object at 0x7efed331d5f0>, '/home/ub/HTM/htm-2d-object-modeling/python/bakedDatabase/pandaVis_dumpData/L2_0_0_proximal.dump'
```

Because enum of the serializable format is not exposed to python, and that i dunno how to do

Reply:
it's possible to add enum to pybind, I've done it for ANOMALY_MODE , should be either in TM or Anomaly bindings. But then again, the current works quite well, so maybe not worth.

Reply:
An app MUST NOT call functions on a region implementations directly but you can use the ExecuteCommand facility in Region to call connections.saveToFile( ).  Calling directly will break the plugin mechanism which needs to work for both Python and C++ on all platforms.  If you are calling saveConnectionsToFile( ) in the python region using the "ExecuteCommand" than this is ok.  I just cannot tell how you are calling it.

By the way, the enums for save( ) and saveToFile( ) are exposed. See htmE\bindings/py/cpp_src/bindings/engine/py_Engine.cpp line 385.   Also, saveToFile( ) without the enum should default to BINARY.   Does that not work?  I also see that the save( ) with the enum was not implemented in the bindings...don't know why.

Reply:
@dkeeney 
> An app MUST NOT call functions on a region implementations directly but you can use the ExecuteCommand facility in Region to call connections.saveToFile( ). Calling directly will break the plugin mechanism which needs to work for both Python and C++ on all platforms. 

I am using executeCommand in Panda, never calling directly, because i treat py regions in same way as cpp regions.

> If you are calling saveConnectionsToFile( ) in the python region using the "ExecuteCommand" than this is ok. I just cannot tell how you are calling it.

I believe that the code responsible for this nice behaviour is here: 
[https://github.com/htm-community/htm.core/blob/56753b0d9ab7e9e17ba837089ee1c8343d80f72f/bindings/py/cpp_src/plugin/PyBindRegion.cpp#L631](https://github.com/htm-community/htm.core/blob/56753b0d9ab7e9e17ba837089ee1c8343d80f72f/bindings/py/cpp_src/plugin/PyBindRegion.cpp#L631)
It basically calls whatever python method and parse result as string.

> By the way, the enums for save( ) and saveToFile( ) are exposed. See htmE\bindings/py/cpp_src/bindings/engine/py_Engine.cpp line 385. Also, saveToFile( ) without the enum should default to BINARY. Does that not work? I also see that the save( ) with the enum was not implemented in the bindings...don't know why.

It reports the error i mentioned above, i must supply serialization type... 
I also don't know. But i not solve this for now, since it is low priority and works like this with save() method.

Reply:
I agree, fixing the enums is low priority. But would you make an issue entry for that so we don't forget about it.  Thanks.

Reply:
Ok created https://github.com/htm-community/htm.core/issues/888